radio/src/gps.cpp

```diff
@@ -104,279 +91,118 @@ uint32_t GPS_coord_to_degrees(const char * coordinateString)
-                gpsData.altitude = gps_Msg.altitude;
-                __enable_irq();
-              }
-              break;
-            case FRAME_RMC:
-              gpsData.speed = gps_Msg.speed;
-              gpsData.groundCourse = gps_Msg.groundCourse;
-#if defined(RTCLOCK)
-              // set RTC clock if needed
-              if (g_eeGeneral.adjustRTC && gps_Msg.fix) {
-                div_t qr = div(gps_Msg.date, 100);
-                uint8_t year = qr.rem;
-                qr = div(qr.quot, 100);
-                uint8_t mon = qr.rem;
-                uint8_t day = qr.quot;
-                qr = div(gps_Msg.time, 100);
-                uint8_t sec = qr.rem;
-                qr = div(qr.quot, 100);
-                uint8_t min = qr.rem;
-                uint8_t hour = qr.quot;
-                rtcAdjust(year+2000, mon, day, hour, min, sec);
-              }
-#endif
-          } // end switch
-        }
-        else {
-          gpsData.errorCount++;
-        }
+      uint32_t new_time = get_tmr10ms();
+      if (new_time - time > 20)  {
+        // No message received
+        changeBaudrate();
+        time = new_time;
       }
-      checksum_param = 0;
-      break;
-    default:
-      if (offset < 15)
-        string[offset++] = c;
-      if (!checksum_param)
-        parity ^= c;
   }
-  return frameOK;
 }
 
-bool gpsNewFrame(uint8_t c)
+static void detectDisconnected(bool has_frame)
 {
-  return gpsNewFrameNMEA(c);
+  static tmr10ms_t time = 0;
+
+  if (has_frame) {
+    time = get_tmr10ms();
+  } else if (time > 0 && get_tmr10ms() - time > 500) {
+    gpsProtocol = GPS_PROTOCOL_AUTO;
+    time = get_tmr10ms();
+  }
 }
 
+
 void gpsNewData(uint8_t c)
 {
-  if (!gpsNewFrame(c)) {
-    return;
+  switch (gpsProtocol) {
+    case GPS_PROTOCOL_NMEA:
+      detectDisconnected(gpsNewFrameNMEA(c));
+      break;
+    case GPS_PROTOCOL_UBX:
+      detectDisconnected(gpsNewFrameUBX(c, false));
+      break;
+    case GPS_PROTOCOL_AUTO:
+      autodetectProtocol(c);
+      break;
   }
 }
 
-static const etx_serial_driver_t* gpsSerialDrv = nullptr;
-static void* gpsSerialCtx = nullptr;
 
 #if defined(DEBUG)
 uint8_t gpsTraceEnabled = false;
 #endif
 
-void gpsSetSerialDriver(void* ctx, const etx_serial_driver_t* drv)
+void gpsSetSerialDriver(void* ctx, const etx_serial_driver_t* drv, int protocol)
 {
+  gpsProtocol = protocol;
   gpsSerialCtx = ctx;
   gpsSerialDrv = drv;
 }
 
 void gpsWakeup()
 {
   if (!gpsSerialDrv) return;
-  
+
   auto _getByte = gpsSerialDrv->getByte;
   if (!_getByte) return;
 
+  static tmr10ms_t time = get_tmr10ms();
   uint8_t byte;
   while (_getByte(gpsSerialCtx, &byte)) {
 #if defined(DEBUG)
     if (gpsTraceEnabled) {
       dbgSerialPutc(byte);
     }
-#endif  
+#endif
     gpsNewData(byte);
+    time = get_tmr10ms();
   }
-}
 
-char hex(uint8_t b) {
-  return b > 9 ? b + 'A' - 10 : b + '0';
+  if (get_tmr10ms() - time > 20) {
+    changeBaudrate();
+    time = get_tmr10ms();
+  }
```

Comment:
I think it might make sense to couple this timeout to the official ACK/NAK timeout of 1s, [page 179 in this receiver spec](https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf). My experience says that this lower timeout is not an issue, but it is not according to the spec

Reply:
Yes, it's a bit aggressive 1s seems a lot though.